core/src/config/stump_config.rs

```diff
@@ -1,191 +1,290 @@
+		let toml_content_str = std::fs::read_to_string(stump_toml)?;
+		let toml_configs = toml::from_str::<PartialStumpConfig>(&toml_content_str)
 			.map_err(|e| CoreError::InitializationError(e.to_string()))?;
 
-		Ok(StumpConfig::from(partial_config))
+		toml_configs.apply_to_config(&mut self);
+		Ok(self)
 	}
 
-	/// Will try to create a new [StumpConfig] object from set environment variables. If none are set,
-	/// will return the default [StumpConfig] object.
-	///
-	/// ## Example
-	/// ```rust
-	/// use std::env;
-	/// use stump_core::config::stump_config::StumpConfig;
-	///
-	/// env::set_var("STUMP_PORT", "8080");
-	/// env::set_var("STUMP_VERBOSITY", "3");
-	///
-	/// let config = StumpConfig::from_env(None);
-	/// assert!(env.is_ok());
-	/// let config = config.unwrap();
-	/// assert_eq!(config.port, 8080);
-	/// assert_eq!(config.verbosity, 3);
-	/// ```
-	pub fn from_env(existing: Option<Self>) -> CoreResult<Self> {
-		let mut config = existing.unwrap_or_default();
-
-		if let Ok(port) = env::var("STUMP_PORT") {
-			config.port = port
-				.parse::<u16>()
-				.map_err(|e| CoreError::InitializationError(e.to_string()))?;
-		}
+	/// Loads configuration variables from the environment, replacing stored
+	/// values with the environment values.
+	pub fn with_environment(mut self) -> CoreResult<Self> {
+		let mut env_configs = PartialStumpConfig::empty();
 
-		if let Ok(profile) = env::var("STUMP_PROFILE") {
+		if let Ok(profile) = env::var(PROFILE_KEY) {
 			if profile == "release" || profile == "debug" {
-				config.profile = profile;
+				env_configs.profile = Some(profile);
 			} else {
 				debug!("Invalid PROFILE value: {}", profile);
-				config.profile = String::from("debug");
 			}
 		}
 
-		if let Ok(verbosity) = env::var("STUMP_VERBOSITY") {
-			config.verbosity = verbosity
+		if let Ok(port) = env::var(PORT_KEY) {
+			let port_u16 = port
+				.parse::<u16>()
+				.map_err(|e| CoreError::InitializationError(e.to_string()))?;
+			env_configs.port = Some(port_u16);
+		}
+
+		if let Ok(verbosity) = env::var(VERBOSITY_KEY) {
+			let verbosity_u64 = verbosity
 				.parse::<u64>()
 				.map_err(|e| CoreError::InitializationError(e.to_string()))?;
+			env_configs.verbosity = Some(verbosity_u64);
 		}
 
-		if let Ok(client_dir) = env::var("STUMP_CLIENT_DIR") {
-			config.client_dir = client_dir;
+		if let Ok(db_path) = env::var(DB_PATH_KEY) {
+			env_configs.db_path = Some(db_path);
 		}
 
-		if let Ok(config_dir) = env::var("STUMP_CONFIG_DIR") {
-			if Path::new(&config_dir).exists() {
-				config.config_dir = config_dir;
-			} else {
-				return Err(CoreError::ConfigDirDoesNotExist(config_dir));
-			}
+		if let Ok(client_dir) = env::var(CLIENT_KEY) {
+			env_configs.client_dir = Some(client_dir);
 		}
 
-		config.config_dir = get_config_dir().to_string_lossy().to_string();
-		config.write()?;
+		if let Ok(config_dir) = env::var(CONFIG_DIR_KEY) {
+			env_configs.config_dir = Some(config_dir);
```

Comment:
I considered this but my thinking was that, because the `write_config_dir()` function will create a non-existing config directory, we might want to support that use case (i.e., the user sets STUMP_CONFIG_DIR to a directory that doesn't exist yet and lets Stump create it for them).

Reply:
Yeah you're totally right, that's a valid scenario